BinaryMarket/contracts/BinaryMarket.sol

```diff
@@ -0,0 +1,710 @@
+import "./interfaces/IOracle.sol";
+
+// fixme I would personally get rid of roundid in oracle. Its redundant information. No need to have it in oracle and I would like if possible to switch oracle to chain link (to allow to write it like we can not that we will do it). That means we can use same oracle for perps. We want to configure perps from existing contracts we have. Less code we need to test and maintain.
+// fixme decide if use error constants or error string, not both. Better to use "ERROR_LIKE_THIS" then "like this"
+contract BinaryMarket is
+    Pausable,
+    IBinaryMarket
+{
+    using SafeERC20 for IERC20;
+
+    // fixme do we real need all fields to be uint256?
+    // legendary, TBH, not sure.. but I think most of them should be.
+    struct Round {
+        uint256 epoch;
+        uint256 startBlock;
+        uint256 lockBlock;
+        uint256 closeBlock;
+        uint256 lockPrice;
+        uint256 closePrice;
+        uint256 lockOracleId;
+        uint256 closeOracleId;
+        uint256 totalAmount;
+        uint256 bullAmount;
+        uint256 bearAmount;
+        bool oracleCalled;
+    }
+
+    struct BetInfo {
+        Position position;
+        uint256 amount;
+        bool claimed; // default false
+    }
+
+    /// @dev Market Data
+    string public marketName;
+    IOracle public oracle;
+    IBinaryVault public vault;
+
+    IERC20 public underlyingToken;
+
+    /// @dev Timeframes supported in this market.
+    TimeFrame[] public timeframes;
+
+    /// @dev Rounds per timeframe
+    mapping(uint8 => mapping(uint256 => Round)) public rounds; // timeframe id => round id => round
+
+    /// @dev bet info per user and round
+    mapping(uint8 => mapping(uint256 => mapping(address => BetInfo)))
+        public ledger; // timeframe id => round id => address => bet info
+
+    // @dev user rounds per timeframe
+    mapping(uint8 => mapping(address => uint256[])) public userRounds; // timeframe id => user address => round ids
+
+    /// @dev current round id per timeframe.
+    mapping(uint8 => uint256) public currentEpochs; // timeframe id => current round id
+
+    /// @dev This should be modified
+    uint256 public minBetAmount;
+    uint256 public oracleLatestRoundId;
+    uint256 public genesisStartBlockTimestamp;
+    uint256 public genesisStartBlockNumber;
+
+    // fixme why admin is not from ownable, why operator is not from accesscontrol?
+    // legendary - We aimed to create all markets from market mananger. 
+    // In this case owner will be market manager, so we cannot control market directly. 
+    // I don't think we need to do all admin operations through market manager. What do you think?
+    address public adminAddress;
+    address public operatorAddress;
+
+    /// @dev default false
+    bool public genesisStartOnce;
+    /// @dev timeframe id => genesis locked? default false
+    mapping(uint8 => bool) public genesisLockOnces;
+
+    event PositionOpened(
```

Comment:
Please order functions layout by type
Refer https://docs.soliditylang.org/en/v0.8.17/style-guide.html#order-of-layout

Recommendation Order:
Events
Modifiers
Constructor
External functions
External functions that are view
External functions that are pure
Public functions
Internal functions
Private functions